### PR TITLE
Ports - Kasey

### PR DESCRIPTION
**Pull request by kaseea**
# JS Adagrams
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
--- | ---
What patterns were you able to use from your Ruby knowledge to apply to JavaScript? | almost everything except hashes. 
Did you need to use different strategies to find information online about JavaScript than you would for Ruby? | yup, MDN was there, but I miss rubydocs
What was a challenge you were able to overcome on this assignment? | not having binding.pry, but got to play around with: https://nodejs.org/api/debugger.html towards the end
What has been interesting and positive about learning a new programming language? | 
What is something to focus on your learnings in JavaScript in the next week? | I'd like to continually use the bits of JavaScript that are different than Ruby, because I think I might just forget what's available and still think like Ruby.

### CODE REVIEW

**Comment by tildeee** [comment]
# JS Adagrams
## What We're Looking For



Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | x
Small commits with meaningful commit messages | x
**Code Requirements** |
`drawLetters` method | x
Uses appropriate data structure to store the letter distribution | x
All tests for `drawLetters` pass | x
`usesAvailableLetters` method | x
All tests for `usesAvailableLetters` pass | x
`scoreWord` method | x
Uses appropriate data structure to store the letter scores | x
All tests for `scoreWord` pass | x
`highestScoreFrom` method | x
Appropriately handles edge cases for tie-breaking logic | x
All tests for `highestScoreFrom` pass | x
**Overall** |



Kasey! Well done on this project! The code looks good, and overall your solutions are great.


I DEFINITELY see what you wanted to do with `highestScoreFrom` and I definitely think there are ways to get there! I haven't put in the time to think through how I would approach it myself, but I'll do that when I can.


Otherwise, I've given you some comments on low-hanging fruit of how I would make the code-style more JavaScript-y.


Overall, great job with this project.

File: src/adagrams.js [hunk 1]
```diff
@@ -1,8 +1,134 @@
-const Adagrams = {
-  drawLetters() {
-    // Implement this method for wave 1
-  },
-};
+class Adagrams  {
+  static drawLetters() {
+      const AllLetters = ['a','a','a','a','a','a','a','a','a','b','b','c','c','d','d','d','d','e','e','e','e','e','e','e','e','e','e','e','e','f','f','g','g','g','h','h','i','i','i','i','i','i','i','i','i','j','k','l','l','l','l','m','m','n','n','n','n','n','n','o','o','o','o','o','o','o','o','p','p','q','r','r','r','r','r','r','s','s','s','s','t','t','t','t','t','t','u','u','u','u','v','v','w','w','x','y','y','z'];
```
Comment (tildeee) [inline src/adagrams.js:3]: in terms of naming convention, I'd either expect this to be `allLetters` or `ALL_LETTERS`

File: src/adagrams.js [hunk 2]
```diff
@@ -1,8 +1,134 @@
...
+      const handy = [];
+      for (let i = 1; i < 11; i++) {
+        let index = [Math.floor(Math.random()*AllLetters.length)];
+        let removed = AllLetters.splice(index, 1);
+        let removed2 = removed.pop()
```
Comment (tildeee) [inline src/adagrams.js:8]: don't forget your semicolon!

File: src/adagrams.js [hunk 3]
```diff
@@ -1,8 +1,134 @@
...
+        handy.push(removed2.toUpperCase());
```
Comment (tildeee) [inline src/adagrams.js:9]: This ends up reading weirdly (having `removed2`), it may make sense to inline some of these definitions. In some ways, I would defend that this line is more readable: `handy.push( removed[0].toUpperCase() );`

File: src/adagrams.js [hunk 4]
```diff
@@ -1,8 +1,134 @@
...
+      }
+      return handy;
+  }
+  static usesAvailableLetters (input, lettersInHand) {
+    let value = true;
```
Comment (tildeee) [inline src/adagrams.js:14]: I would probably prefer a different variable name, as `value` doesn't really mean anything to me in terms of understanding the code

File: src/adagrams.js [hunk 5]
```diff
@@ -1,8 +1,134 @@
...
+    [...input].forEach(letter => {
```
Comment (tildeee) [inline src/adagrams.js:15]: Nice use of the spread operator!

File: src/adagrams.js [hunk 6]
```diff
@@ -1,8 +1,134 @@
...
+      // ugh forEach CAN"T BE ESCAPED WITH A RETURN OR BREAK
+      // for(x in y) use in for objects
+      // for(x of y) for strings or arrays or anything else
+      let index = lettersInHand.indexOf(letter);
```
Comment (tildeee) [inline src/adagrams.js:19]: `index` can be a `const`

File: src/adagrams.js [hunk 7]
```diff
@@ -1,8 +1,134 @@
...
+      'r': 1,
+      's': 1,
+      't': 1,
+      'u': 1,
+      'v': 4,
+      'w': 4,
+      'x': 8,
+      'y': 4,
+      'z': 10,
+    }
```
Comment (tildeee) [inline src/adagrams.js:56]: JS defaults keys to strings, so turns out you can just do `const tile = { a: 1, b: 3, c: 3 ...}`

File: src/adagrams.js [hunk 8]
```diff
@@ -1,8 +1,134 @@
...
+
+    // so I know I'm updating score, but I"M JUST GOING TO UPDATE IT so shouldn't I be allowed to use a const?
```
Comment (tildeee) [inline src/adagrams.js:58]: For variables that hold integers, "modifying" means re-assignment, and `const` doesn't allow for re-assignment

File: src/adagrams.js [hunk 9]
```diff
@@ -1,8 +1,134 @@
...
+    let score = 0;
+
+    [...word].forEach(letter => {
+      letter = letter.toLowerCase();
+      // WHY CAN"T I USE tile.letter?
+      // Cause its a string 'b' instead of b right?
+      // is there a good work around?
```
Comment (tildeee) [inline src/adagrams.js:65]: Yes! it's because `letter` is a variable. dot notation doesn't work with variables, it is looking for a member named whatever that thing is (aka `letter`). `[]` notation is necessary for this kind of thing

File: src/adagrams.js [hunk 10]
```diff
@@ -1,8 +1,134 @@
...
+      score += tile[letter];
+    });
+
+    if (word.length === 7 || word.length === 8 || word.length === 9 || word.length === 10 ) {
```
Comment (tildeee) [inline src/adagrams.js:69]: Instead of checking if `word.length` IS any of these values, I may just go for `word.length >= 7`

File: src/adagrams.js [hunk 11]
```diff
@@ -1,8 +1,134 @@
...
+      score += 8
```
Comment (tildeee) [inline src/adagrams.js:70]: semicolon

File: src/adagrams.js [hunk 12]
```diff
@@ -1,8 +1,134 @@
...
+      // but with an accumulator I can only figure it out how to get it like {
+      //   a: { word: 'a', score: 1 },
+      //   dog: { word: 'dog', score: 5 },
+      //   god: { word: 'god', score: 5 }
+      // }
+      // WHICH CREATES ALL THE USELESS CODE ALL THE WAY DOWN
+        // allScores[word] = this.scoreWord(word);
+        // below creates { word: 'god', score: 5 } but only one cause the accumulator gets overwritten?
+        // allScores = { 'word': word, 'score': this.scoreWord(word) };
+        // return allScores;
```
Comment (tildeee) [inline src/adagrams.js:91]: There is a way to make this work! This sounds like a painful journey though hahahaha.

File: src/adagrams.js [hunk 13]
```diff
@@ -1,8 +1,134 @@
...
+        let score = this.scoreWord(word);
```
Comment (tildeee) [inline src/adagrams.js:92]: `score` can be a `const`!

File: src/adagrams.js [hunk 14]
```diff
@@ -1,8 +1,134 @@
...
+        if (score > max) {
+          max = score;
+        }
+        allScores.push({ 'word': word, 'score': this.scoreWord(word) });
```
Comment (tildeee) [inline src/adagrams.js:96]: consider `allScores.push({ word: word, score: score });` or better yet `allScores.push({ word, score });`
